implementations/rust/ockam/ockam_command/src/util/installer.rs

```diff
@@ -0,0 +1,411 @@
+use crate::error::Result;
+use miette::miette;
+use miette::Context;
+use miette::IntoDiagnostic;
+use std::fs;
+use std::fs::Permissions;
+use std::path::Path;
+use std::process::Stdio;
+use std::{path::PathBuf, process::Command};
+
+use crate::util::github::{download_install_file_sync, download_latest_binary_sync};
+
+/// Get the appropriate installer to upgrade and uninstall ockam.
+pub fn get_installer() -> Result<Box<dyn Installer>> {
+    if installed_with_brew()? {
+        return Ok(Box::<BrewInstaller>::default());
+    } else if installed_with_script()? {
+        return Ok(Box::<ScriptInstaller>::default());
+    }
+    Ok(Box::<BinaryInstaller>::default())
+}
+
+pub trait Installer {
+    /// Uninstall ockam.
+    fn uninstall(&self) -> miette::Result<()>;
+    /// Upgrade ockam to the latest version.
+    fn upgrade(&self) -> miette::Result<()>;
+}
+
+fn installed_with_brew() -> Result<bool> {
+    // First we check to see if the ockam binary matches the brew list path
+    // We check the canonical file paths to ensure that all paths are absoulte
+    // and the simlinks are resolved.
+    // Any unexpected errors here should cancel the upgrade/uninstallation process
+    // as we don't want to accidentally delete the wrong files.
+    let brew_prefix = Command::new("brew").args(["--prefix", "ockam"]).output();
+    match brew_prefix {
+        Ok(output) => {
+            if output.status.success() {
+                let brew_prefix_str = String::from_utf8(output.stdout).into_diagnostic()?;
+                let brew_ockam_path = PathBuf::from(brew_prefix_str.trim()).join("bin/ockam");
+                let canonical_brew_ockam_path =
+                    fs::canonicalize(brew_ockam_path).into_diagnostic()?;
+                let current_exe = std::env::current_exe().into_diagnostic()?;
+                let canonical_current_exe = fs::canonicalize(current_exe).into_diagnostic()?;
+                if canonical_brew_ockam_path == canonical_current_exe {
+                    return Ok(true);
+                }
+            }
+            Ok(false)
+        }
+        Err(_) => Ok(false),
```

Comment:
It would be good to print the return message of the `brew` command in case of an error.

Reply:
I've added an `eprintln!` statement for this

Reply:
Can you please use `opts.terminal` in that case too?

Reply:
Done! Let me know what you think. Will look like this for the upgrade check.
<img width="872" alt="image" src="https://github.com/build-trust/ockam/assets/22850972/1095bb9f-6514-4c23-b5e1-2627c0639d60">